lumigator/python/mzai/backend/backend/services/workflows.py

```diff
@@ -0,0 +1,166 @@
+import asyncio
+import datetime
+import uuid
+from collections.abc import Callable
+from uuid import UUID
+
+import loguru
+from fastapi import BackgroundTasks
+from lumigator_schemas.jobs import (
+    JobEvalLiteCreate,
+    JobInferenceCreate,
+    JobStatus,
+)
+from lumigator_schemas.workflows import WorkflowCreate, WorkflowResponse
+
+from backend.services.datasets import DatasetService
+from backend.services.jobs import JobService
+
+
+class WorkflowService:
+    def __init__(
+        self,
+        job_service: JobService,
+        dataset_service: DatasetService,
+    ):
+        self._job_service = job_service
+        self._dataset_service = dataset_service
+
+    NON_TERMINAL_STATUS = [
+        JobStatus.CREATED.value,
+        JobStatus.PENDING.value,
+        JobStatus.RUNNING.value,
+    ]
+    """list: A list of non-terminal job statuses."""
+
+    # TODO: rely on https://github.com/ray-project/ray/blob/7c2a200ef84f17418666dad43017a82f782596a3/python/ray/dashboard/modules/job/common.py#L53
+    TERMINAL_STATUS = [JobStatus.FAILED.value, JobStatus.SUCCEEDED.value]
+    """list: A list of terminal job statuses."""
+
+    async def on_job_complete(self, job_id: UUID, task: Callable = None, *args):
+        """Watches a submitted job and, when it terminates successfully, runs a given task.
+
+        Inputs:
+        - job_id: the UUID of the job to watch
+        - task: the function to be called after the job completes successfully
+        - args: the arguments to be passed to the function `task()`
+        """
+        job_status = self._job_service.get_upstream_job_status(job_id)
+
+        loguru.logger.info(f"Watching {job_id}")
+        while job_status not in self.TERMINAL_STATUS and job_status in self.NON_TERMINAL_STATUS:
+            await asyncio.sleep(5)
+            job_status = self._job_service.get_upstream_job_status(job_id)
+
+        match job_status:
+            case JobStatus.FAILED.value:
+                loguru.logger.error(f"Job {job_id} failed: not running task {str(task)}")
+            case JobStatus.SUCCEEDED.value:
+                loguru.logger.info(f"Job {job_id} finished successfully.")
+                if task is not None:
+                    task(*args)
```

Comment:
would be helpful to leave a comment here understanding what args could potentially be passed, I'm always wary of args and kwargs in python since they lead to tricky debugging in the case of collected elements 😅

Reply:
Totally! This is existing code and I also have my eye on refactoring it, but would like to defer this work to a later PR.